### PR TITLE
feat(*): refactor error handling to preserve stack traces on platform exceptions

### PR DESCRIPTION
**Pull request by rrousselGit**
## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [ ] All existing and new tests are passing.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] The analyzer (`melos run analyze`) does not report any problems on my PR.
- [ ] I read and followed the [Flutter Style Guide].
- [ ] I signed the [CLA].
- [ ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change.
- [ ] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/FirebaseExtended/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Review by Salakar — APPROVED** [review]

**Comment by ben-xx** [comment]
@rrousselGit Hey Rémi, should min Dart SDK version be moved from 2.12 to 2.16 [on pub.dev](https://pub.dev/packages/firebase_messaging/versions) with this change?

https://api.flutter.dev/flutter/dart-core/Error/throwWithStackTrace.html was added in Dart 2.16 I believe.

After updating `firebase_messaging_platform_interface` from  3.1.3 to latest 3.2.0 I started getting:
```
/C:/Users/Blah/AppData/Local/Pub/Cache/hosted/pub.dartlang.org/firebase_messaging_platform_interface-3.2.0/lib/src/method_channel/utils/exception.dart:13:11: Error: Member not found: 'Error.throwWithStackTrace'.
    Error.throwWithStackTrace(exception, stackTrace);
          ^^^^^^^^^^^^^^^^^^^
```

Moving back down to 3.1.3 in pubspec.yaml avoided the above:
```
dependency_overrides:
  firebase_messaging_platform_interface: 3.1.3
```

**Comment by dreamer2q** [comment]
This changes has broken people whose Dart SDK < 2.16 and make people disappointed.

**Comment by rrousselGit** [comment]
The pubpsec  of the changed package was correctly updated to mention >= 2.16 (see https://pub.dev/packages/firebase_core/versions)

It seems like pub isn't picking up transitive dependencies constraints thought

We can update **all** packages to explicitly require 2.16 (even if they weren't directly changed by this PR) if needed. Although it's not going to change the fact that you'd need Dart >= 2.16

**Comment by ben-xx** [comment]
@rrousselGit Awesome! Thanks Rémi for the update & work on firebase changes / new Error API.  Much appreciated.

**Comment by jpetro416** [comment]
What is going on?!!! Why would you force us who are using less than 2.16 to get screwed over. WTF

File: packages/firebase_core/firebase_core/pubspec.yaml [hunk 1]
```diff
@@ -9,7 +9,7 @@ false_secrets:
   - example/**
 
 environment:
-  sdk: ">=2.12.0 <3.0.0"
+  sdk: ">=2.16.0 <3.0.0"
```
[the remark this answers is not in the document]
Comment (rrousselGit) [reply packages/firebase_core/firebase_core/pubspec.yaml:12]: pub should handle transitive dependencies too.

[the remark this answers is not in the document]
Comment (dreamer2q) [reply packages/firebase_core/firebase_core/pubspec.yaml:12]: > 

As far as I know, flutter is fetching packages from pub.dev and these packages once released are snapshots (which means they are unmodified). The newly released packages still rely on the old published firebase_core (which require sdk > 2.12) and the packages themselves still require sdk > 2.12, so there no change in sdk requirement but the actual code uses feature provided by sdk 2.16.

In short, you cannot expect a released package to upgrade its min sdk requirement, do you?

[the remark this answers is not in the document]
Comment (rrousselGit) [reply packages/firebase_core/firebase_core/pubspec.yaml:12]: Published packages don't have to update their sdk requirement.

It's about pub checking transitive dependencies on installation to verify that they all respect sdk constraints.

**Pull request merged**